File: apps/android/src/main/java/com/tilawah/android/features/profile/PreferencesViewModel.kt
package com.tilawah.android.features.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tilawah.android.backend.PracticePreferences
import com.tilawah.android.backend.PracticePreferencesPatch
import com.tilawah.android.backend.PreferencesClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Drives the Practice Preferences card with an optimistic-write model:
 * each edit applies locally at once, fires a single-field `PATCH`, and
 * rolls back with an error message if the server rejects it. The server's
 * echo — not the local guess — becomes the source of truth so a
 * server-side normalisation (e.g. `1` → `1.0`) sticks.
 *
 * Mirrors `apps/ios/.../Features/Profile/PreferencesViewModel.swift`.
 */
class PreferencesViewModel(
    private val client: PreferencesClient,
) : ViewModel() {

    private val _state = MutableStateFlow(
        PreferencesUiState(
            preferences = PracticePreferences.Default,
            isLoaded = false,
            errorMessage = null,
        ),
    )
    val state: StateFlow<PreferencesUiState> = _state.asStateFlow()

    /**
     * Loads once per instance. A failed load keeps the defaults rather
     * than blocking the card — "user has no row yet" is the common case
     * and the BFF returns defaults for it, so a transient fetch error
     * degrades to the same place.
     */
    fun load() {
        if (_state.value.isLoaded) return
        viewModelScope.launch {
            val loaded = try {
                client.preferences()
            } catch (_: Throwable) {
                null
            }
            _state.update { it.copy(preferences = loaded ?: it.preferences, isLoaded = true) }
        }
    }

    fun setReciter(id: String) =
        applyEdit(PracticePreferencesPatch(referenceReciterId = id)) { it.copy(referenceReciterId = id) }

    fun setPlaybackSpeed(speed: Double) =
        applyEdit(PracticePreferencesPatch(defaultPlaybackSpeed = speed)) { it.copy(defaultPlaybackSpeed = speed) }

    fun setReminderEnabled(enabled: Boolean) =
        applyEdit(PracticePreferencesPatch(dailyReminderEnabled = enabled)) { it.copy(dailyReminderEnabled = enabled) }

    fun setReminderTime(time: String) =
        applyEdit(PracticePreferencesPatch(dailyReminderTime = time)) { it.copy(dailyReminderTime = time) }

    /**
     * Apply [optimistic] locally, PATCH [patch], and reconcile against the
     * server echo — or roll back to the pre-edit snapshot on failure.
     */
    private fun applyEdit(
        patch: PracticePreferencesPatch,
        optimistic: (PracticePreferences) -> PracticePreferences,
    ) {
        val previous = _state.value.preferences
        _state.update { it.copy(preferences = optimistic(it.preferences), errorMessage = null) }
        viewModelScope.launch {
            try {
                val echoed = client.updatePreferences(patch)
                _state.update { it.copy(preferences = echoed) }
            } catch (_: Throwable) {
                _state.update { it.copy(preferences = previous, errorMessage = SAVE_FAILED_MESSAGE) }
            }
        }
    }

    private companion object {
        const val SAVE_FAILED_MESSAGE = "Couldn't save preference. Try again."
    }
}

/** Snapshot the Practice Preferences card observes. */
data class PreferencesUiState(
    val preferences: PracticePreferences,
    /** Whether the first GET has resolved — gates the loading affordance. */
    val isLoaded: Boolean,
    val errorMessage: String?,
)

File: apps/ios/Sources/QuranRecitationApp/Features/Profile/PreferencesViewModel.swift
import Foundation

/// Drives the Practice Preferences card with an optimistic-write model:
/// each edit applies locally at once, fires a single-field `PATCH`, and
/// rolls back with an error banner if the server rejects it.
@MainActor
final class PreferencesViewModel: ObservableObject {
  @Published private(set) var preferences: PracticePreferences
  /// Tracks whether the first GET has resolved so the card can show a
  /// spinner instead of flashing the optimistic defaults as if loaded.
  @Published private(set) var isLoaded = false
  @Published var errorMessage: String?

  private let meClient: MeClient

  init(meClient: MeClient, preferences: PracticePreferences = .default) {
    self.meClient = meClient
    self.preferences = preferences
  }

  /// Loads once per instance. A failed load keeps the defaults rather
  /// than blocking the card — the user has no row yet is the common
  /// case, and the BFF itself returns defaults for that, so a transient
  /// fetch error degrades to the same place.
  func loadIfNeeded() async {
    guard !isLoaded else { return }
    if let loaded = try? await meClient.preferences() {
      preferences = loaded
    }
    isLoaded = true
  }

  func setReciter(_ id: String) async {
    await apply(PracticePreferencesPatch(referenceReciterId: id)) { $0.referenceReciterId = id }
  }

  func setPlaybackSpeed(_ speed: Double) async {
    await apply(PracticePreferencesPatch(defaultPlaybackSpeed: speed)) { $0.defaultPlaybackSpeed = speed }
  }

  func setReminderEnabled(_ isOn: Bool) async {
    await apply(PracticePreferencesPatch(dailyReminderEnabled: isOn)) { $0.dailyReminderEnabled = isOn }
  }

  func setReminderTime(_ time: String) async {
    await apply(PracticePreferencesPatch(dailyReminderTime: time)) { $0.dailyReminderTime = time }
  }

  /// Apply `optimistic` locally, PATCH `patch`, and reconcile against
  /// the server's echo — or roll back to the pre-edit snapshot on
  /// failure. The echo (not the local guess) becomes the source of
  /// truth so a server-side normalisation (e.g. `1` → `1.00`) sticks.
  private func apply(
    _ patch: PracticePreferencesPatch,
    _ optimistic: (inout PracticePreferences) -> Void
  ) async {
    let previous = preferences
    optimistic(&preferences)
    errorMessage = nil
    do {
      preferences = try await meClient.updatePreferences(patch)
    } catch {
      preferences = previous
      errorMessage = Self.saveFailedMessage
    }
  }

  static let saveFailedMessage = "Couldn't save preference. Try again."
}
